redis/server.cc

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright (C) 2019 pengjian.uestc @ gmail.com
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "redis/server.hh"
+#include "service/storage_service.hh"
+#include "db/consistency_level_type.hh"
+#include "db/config.hh"
+#include "db/write_type.hh"
+#include <seastar/core/future-util.hh>
+#include <seastar/core/reactor.hh>
+#include <seastar/core/metrics.hh>
+#include <seastar/net/byteorder.hh>
+#include <seastar/core/execution_stage.hh>
+#include "service/query_state.hh"
+#include "exceptions/exceptions.hh"
+#include "auth/authenticator.hh"
+#include <cassert>
+#include <string>
+#include "redis/request.hh"
+#include "redis/reply.hh"
+#include <unordered_map>
+
+namespace redis_transport {
+
+static logging::logger logging("redis_server");
+
+redis_server::redis_server(distributed<service::storage_proxy>& proxy, distributed<redis::query_processor>& qp, auth::service& auth_service, redis_server_config config)
+    : _proxy(proxy)
+    , _query_processor(qp)
+    , _config(config)
+    , _max_request_size(config._max_request_size)
+    , _memory_available(_max_request_size)
+    , _auth_service(auth_service)
+{
+}
+
+future<> redis_server::stop() {
+    _stopping = true;
+    size_t nr = 0;
+    size_t nr_total = _listeners.size();
+    logging.debug("redis_server: abort accept nr_total={}", nr_total);
+    for (auto&& l : _listeners) {
+        l.abort_accept();
+        logging.debug("redis_server: abort accept {} out of {} done", ++nr, nr_total);
+    }
+    auto nr_conn = make_lw_shared<size_t>(0);
+    auto nr_conn_total = _connections_list.size();
+    logging.debug("redis_server: shutdown connection nr_total={}", nr_conn_total);
+    return parallel_for_each(_connections_list.begin(), _connections_list.end(), [nr_conn, nr_conn_total] (auto&& c) {
```

Comment:
We can fix this later, but it's probably not a great idea to do something in parallel for a a hue number of connections